File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第9章：几何形状的检测和拟合/9.2/HTLine.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
#include<map>
#include<iostream>
using namespace std;
map<vector<int>, vector<Point>> HTLine(Mat I,Mat & accumulator,float stepTheta=1, float stepRho=1)
{
	//图像的宽高
	int rows = I.rows;
	int cols = I.cols;
	//可能出现的最大垂线的长度
	int L = round(sqrt(pow(rows-1,2.0)+pow(cols-1,2.0)))+1;
	//初始化投票器
	int numtheta = int(180.0 / stepTheta);
	int numRho = int(2 * L / stepRho + 1);
	accumulator=Mat::zeros(Size(numtheta, numRho), CV_32SC1);

	//初始化 map 类，用于存储共线的点
	map<vector<int> , vector<Point>> lines;
	for (int i = 0; i < numRho; i++)
	{
		for (int j = 0; j < numtheta; j++)
		{
			lines.insert(make_pair(vector<int>(j,i),vector<Point>()));
		}
	}
	//投票计数
	for (int y = 0; y < rows; y++)
	{
		for (int x = 0; x < cols; x++)
		{
			if (I.at<uchar>(Point(x, y)) == 255)
			{
				for (int m = 0; m < numtheta; m++)
				{
					//对每一个角度，计算对应的 rho 值
					float rho1 = x*cos(stepTheta*m / 180.0*CV_PI);
					float rho2 = y*sin(stepTheta*m / 180.0*CV_PI);
					float rho = rho1 + rho2;
					//计算投票到哪一个区域
					int n = int(round(rho + L) / stepRho);
					//累加 1
					accumulator.at<int>(n, m) += 1;
					//记录该点
					lines.at(vector<int>(m,n)).push_back(Point(x,y));
				}
			}
		}
	}
	return lines;
}
int main(int argc, char*argv[])
{
	//输入图像
	Mat img = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	//图像边缘检测
	Mat edge;
	Canny(img, edge, 50, 200);
	//霍夫直线检测
	Mat accu;//投票器
	map<vector<int>, vector<Point>> lines;
	lines = HTLine(edge, accu);
	//投票器的灰度级可视化
	double maxValue;//找到投票器中的最大值
	minMaxLoc(accu, NULL, &maxValue, NULL, NULL);
	//数据类型转换，对投票器进行灰度化
	Mat grayAccu;
	accu.convertTo(grayAccu, CV_8UC1,1.0 / maxValue*255);
	grayAccu = grayAccu.t();
	imshow("投票器的灰度级显示", grayAccu);
	imwrite("grayAccu.jpg", grayAccu);
	//画出投票数大于某一阈值的直线
	int vote = 150;
	for (int r = 1; r < accu.rows-1; r++)
	{
		for (int c = 1; c < accu.cols-1; c++)
		{	
			int current = accu.at<int>(r, c);
			//画直线：line 的首末元素为起始点
			if (current> vote)
			{
				
				int lt = accu.at<int>(r - 1, c - 1);//左上
				int t = accu.at<int>(r - 1, c );//正上
				int rt = accu.at<int>(r - 1, c+1);//右上
				int l = accu.at<int>(r, c - 1);//左
				int right = accu.at<int>(r, c + 1);//右
				int lb = accu.at<int>(r+1, c - 1);//左下
				int b = accu.at<int>(r - 1, c );//下
				int rb = accu.at<int>(r + 1, c+1);//下
				 //判断该位置是不是局部最大值
				if (current > lt&& current > t && current > rt
					&&current > l && current > right&&
					current > lb && current > b&& current > rb)
				{
					vector<Point> line = lines.at(vector<int>(c, r));
					int s = line.size();
					//画线
					cv::line(img, line.at(0), line.at(s - 1), Scalar(255), 2);
				}	
			}
		}
	}
	imshow("img", img);
	waitKey(0);
}

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第9章：几何形状的检测和拟合/9.2/HTLine.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
import cv2
import math
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
#霍夫极坐标变换：直线检测
def HTLine (image,stepTheta=1,stepRho=1):
    #宽、高
    rows,cols = image.shape
    #图像中可能出现的最大垂线的长度
    L =  round(math.sqrt(pow(rows-1,2.0)+pow(cols-1,2.0)))+1
    #初始化投票器
    numtheta = int(180.0/stepTheta)
    numRho = int(2*L/stepRho + 1)
    accumulator = np.zeros((numRho,numtheta),np.int32)
    #建立字典
    accuDict={}
    for k1 in xrange(numRho):
        for k2 in xrange(numtheta):
            accuDict[(k1,k2)]=[]
    #投票计数
    for y in xrange(rows):
        for x  in xrange(cols):
            if(image[y][x] == 255):#只对边缘点做霍夫变换
                for m in xrange(numtheta):
                    #对每一个角度，计算对应的 rho 值
                    rho = x*math.cos(stepTheta*m/180.0*math.pi)+y*math.sin(stepTheta*m/180.0*math.pi)
                    #计算投票哪一个区域
                    n = int(round(rho+L)/stepRho)
                    #投票加 1
                    accumulator[n,m] += 1
                    #记录该点
                    accuDict[(n,m)].append((x,y))
    return accumulator,accuDict
#主函数
if __name__ == "__main__":
    if len(sys.argv)>1:
        #输入图像
        I = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usage: python HTPLine.py image"
    #canny 边缘检测
    edge = cv2.Canny(I,50,200)
    #显示二值化边缘
    cv2.imshow("edge",edge)
    #霍夫直线检测
    accumulator,accuDict = HTLine(edge,1,1)
    #计数器的二维直方图方式显示
    rows,cols = accumulator.shape
    fig = plt.figure()
    ax = fig.gca(projection='3d')
    X,Y = np.mgrid[0:rows:1, 0:cols:1]
    surf = ax.plot_wireframe(X,Y,accumulator,cstride=1, rstride=1,color='gray')
    ax.set_xlabel(u"$\\rho$")
    ax.set_ylabel(u"$\\theta$")
    ax.set_zlabel("accumulator")
    ax.set_zlim3d(0,np.max(accumulator))
    #计数器的灰度级显示
    grayAccu = accumulator/float(np.max(accumulator))
    grayAccu = 255*grayAccu
    grayAccu = grayAccu.astype(np.uint8)
    #只画出投票数大于 60 直线
    voteThresh = 60
    for r in xrange(rows):
        for c in xrange(cols):
            if accumulator[r][c] > voteThresh:
                points = accuDict[(r,c)]
                cv2.line(I,points[0],points[len(points)-1],(255),2)
    cv2.imshow('accumulator',grayAccu)
    
    #显示原图
    cv2.imshow("I",I)
    plt.show()
    cv2.imwrite('accumulator.jpg',grayAccu)
    cv2.imwrite('I.jpg',I)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
